### PR TITLE
search addon migration

### PR DESCRIPTION
**Pull request by zhiweiyin318**
story : https://github.com/stolostron/backlog/issues/18488

there are 3 PRs for search addon migration.
1.  https://github.com/stolostron/search-chart/pull/132 -- change the rbac and operator deployment.
2. this pr  -- add addon installation using addon-framework.
3. https://github.com/stolostron/klusterlet-addon-controller/pull/95 -- disable old search addon and enable new one.

please merge 1 firstly, and merge 3 at the last .
Signed-off-by: Zhiwei Yin <zyin@redhat.com>

### CODE REVIEW

File: main.go
Comment (jlpadilla) [inline main.go:94]: I don't want to exit on these errors because this operator does other functions.  These errors create a significant degradation, but we could still perform the other tasks.

File: config/crd/ocm/addon.open-cluster-management.io_clustermanagementaddons.crd.yaml
Comment (jlpadilla) [inline config/crd/ocm/addon.open-cluster-management.io_clustermanagementaddons.crd.yaml:2]: Why do we need these CRDs?  These are common and I'd expect the add-on framework to install these.
If each add-on tries to install these, it will create problems, especially if upgrading or deleting the helm release.

Comment (zhiweiyin318) [reply config/crd/ocm/addon.open-cluster-management.io_clustermanagementaddons.crd.yaml:2]: these CRDs are install by OCM, in e2e tests,  OCM is not deployed, so need to install them for e2e tests.

Comment (zhiweiyin318) [reply config/crd/ocm/addon.open-cluster-management.io_clustermanagementaddons.crd.yaml:2]: they are deployed here in run-search-operator-e2e.sh https://github.com/stolostron/search-operator/pull/99/files#diff-ff7c1281606eabd00ba3bbc58b1bef5e6c121164394b948c7371071e684f100dR144 for e2e tests.

File: main.go
Comment (zhiweiyin318) [reply main.go]: done

File: addon/manifests/chart/templates/role_binding.yaml
Comment (jlpadilla) [inline addon/manifests/chart/templates/role_binding.yaml:5]: Where does this RoleBinding gets used? I assume this get deployed to the managed cluster, but the RoleBinging for the search-collector is defined in a different file (above). Is there any other pod that uses this one?

File: config/crd/ocm/addon.open-cluster-management.io_clustermanagementaddons.crd.yaml
Comment (jlpadilla) [reply config/crd/ocm/addon.open-cluster-management.io_clustermanagementaddons.crd.yaml:2]: Thanks! I thought these were going to the managed cluster.  It's good if it's only for the tests.

File: main.go
Comment (jlpadilla) [inline main.go:92]: Is it possible to recover from this error? I wonder if there's any benefit from adding a retry logic.

File: main.go
Comment (jlpadilla) [reply main.go]: thanks!

**Comment by jlpadilla** [comment]
I have 2 minor questions open, otherwise changes look good.
I want to get @xrajesh 's review before merging.

File: addon/manifests/chart/templates/role_binding.yaml
Comment (zhiweiyin318) [reply addon/manifests/chart/templates/role_binding.yaml:5]: I copied the full chart from klusterlet-addon-operator https://github.com/stolostron/klusterlet-addon-operator/tree/main/helm-charts/search/templates 
I have the same question too.  if search-collector do not need get/delete pod permission, I will delete it. let me check if there is some historical reason with the previous maintainer.

File: main.go
Comment (zhiweiyin318) [reply main.go:92]: I don't think so. `addonMgr` is a controller,  and `addon.NewAddonManager` only does some static logics.  I think it is impossible to recover if there is error.
so could we consider to` os.Exit(1)` here ?  because search-collector-addon as a part of search-operator, the whole search feature cannot work if the addonMgr does not start working.

File: addon/manifests/chart/templates/role_binding.yaml
Comment (jlpadilla) [reply addon/manifests/chart/templates/role_binding.yaml:5]: Get permission is already granted by `cluster_role.yaml` and we don't need the delete pod permission. So it's ok to remove this and `role.yaml`.

File: main.go
Comment (jlpadilla) [reply main.go:92]: If we hit this error, search could still work, but it will be degraded to only resources in the hub. If we exit, then the hub components won't get deployed, which causes a worse problem.

Both scenarios aren't good, but exiting will cause a more severe problem. On the other hand, continuing could make it harder to notice and debug the problem. Could we show this error in the operator's status?

@xrajesh What are your thoughts for this?

File: addon/manifests/chart/templates/role_binding.yaml
Comment (zhiweiyin318) [reply addon/manifests/chart/templates/role_binding.yaml]: removed.

Comment (jlpadilla) [reply addon/manifests/chart/templates/role_binding.yaml]: thanks!

File: main.go
Comment (xrajesh) [reply main.go:92]: @zhiweiyin318 We can put a log message when the addon is successfully created - Helps if there is a problem, because we are quite when everything is working .

**Review by xrajesh — APPROVED** [review]

**Pull request merged**
